src/python/pants/engine/goal.py

```diff
@@ -133,6 +135,27 @@ def name(cls) -> str:
         return cast(str, cls.subsystem_cls.name)
 
 
+@dataclass(frozen=True)
+class CurrentExecutingGoal(EngineAwareReturnType):
```

Comment:
Made this a "singleton" instance, but mutable and uncachable, to make sure that the `self.goal` value is up-to-date at the time a rule is looking at it. Hopefully 😁